src/compiler/build_tables/build_parse_table.cc

```diff
@@ -394,7 +394,8 @@ class ParseTableBuilder {
       } else {
         ParseState &state = *iter;
         state.each_referenced_state([&new_state_ids](ParseStateId *state_index) {
-          *state_index = new_state_ids[*state_index];
+          if (*state_index != (ParseStateId)(-1))
+            *state_index = new_state_ids[*state_index];
```

Comment:
My intention was to for `each_referenced_state` to only call this callback with valid state indices, so we shouldn't have to check them here.

I think all we need is your other fix to `each_referenced_state` itself, plus the one additional check I mention below.